File: frontend/src/api/backup.ts
import apiClient from './client'

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface BackupDestination {
  id: number
  name: string
  type: 'local' | 'smb' | 'onedrive'
  enabled: boolean
  rclone_remote_name: string
  rclone_config: string | null
  target_path: string
  created_at: string
}

export interface BackupDestinationCreate {
  name: string
  type: 'local' | 'smb' | 'onedrive'
  enabled?: boolean
  rclone_remote_name: string
  rclone_config?: string | null
  target_path?: string
}

export interface BackupDestinationUpdate {
  name?: string
  type?: 'local' | 'smb' | 'onedrive'
  enabled?: boolean
  rclone_remote_name?: string
  rclone_config?: string | null
  target_path?: string
}

export interface BackupSchedule {
  enabled: boolean
  schedule_type: 'interval' | 'cron'
  interval_hours: number
  cron_expression: string | null
  include_uploads: boolean
  notify_on_failure: boolean
}

export interface BackupScheduleUpdate {
  enabled?: boolean
  schedule_type?: 'interval' | 'cron'
  interval_hours?: number
  cron_expression?: string | null
  include_uploads?: boolean
  notify_on_failure?: boolean
}

export interface BackupRunStatus {
  last_run_at: string | null
  last_run_status: 'success' | 'failure' | null
  destinations_results: Array<{
    destination_id: number
    destination_name: string
    success: boolean
    error: string | null
  }>
}

export interface BackupConnectionTestResult {
  success: boolean
  message: string
}

export interface BackupRestoreTestResult {
  ok: boolean
  integrity_check: string
  tables_found: string[]
  tables_missing: string[]
  error: string | null
}

export interface OneDriveOAuthStart {
  port: number
  auth_url: string
}

export interface OneDriveOAuthStatus {
  done: boolean
  token: string | null
}

// ---------------------------------------------------------------------------
// API calls — destinations
// ---------------------------------------------------------------------------

export async function listDestinations(): Promise<BackupDestination[]> {
  const { data } = await apiClient.get<BackupDestination[]>('/api/backup/destinations')
  return data
}

export async function createDestination(
  payload: BackupDestinationCreate,
): Promise<BackupDestination> {
  const { data } = await apiClient.post<BackupDestination>('/api/backup/destinations', payload)
  return data
}

export async function updateDestination(
  id: number,
  payload: BackupDestinationUpdate,
): Promise<BackupDestination> {
  const { data } = await apiClient.put<BackupDestination>(`/api/backup/destinations/${id}`, payload)
  return data
}

export async function deleteDestination(id: number): Promise<void> {
  await apiClient.delete(`/api/backup/destinations/${id}`)
}

export async function testDestination(id: number): Promise<BackupConnectionTestResult> {
  const { data } = await apiClient.post<BackupConnectionTestResult>(
    `/api/backup/destinations/${id}/test`,
  )
  return data
}

// ---------------------------------------------------------------------------
// API calls — schedule
// ---------------------------------------------------------------------------

export async function getSchedule(): Promise<BackupSchedule> {
  const { data } = await apiClient.get<BackupSchedule>('/api/backup/schedule')
  return data
}

export async function updateSchedule(payload: BackupScheduleUpdate): Promise<BackupSchedule> {
  const { data } = await apiClient.put<BackupSchedule>('/api/backup/schedule', payload)
  return data
}

// ---------------------------------------------------------------------------
// API calls — run + status
// ---------------------------------------------------------------------------

export async function triggerBackup(): Promise<void> {
  await apiClient.post('/api/backup/run')
}

export async function getBackupStatus(): Promise<BackupRunStatus> {
  const { data } = await apiClient.get<BackupRunStatus>('/api/backup/status')
  return data
}

// ---------------------------------------------------------------------------
// API calls — test-restore + restore
// ---------------------------------------------------------------------------

export async function testRestoreBackup(filename: string): Promise<BackupRestoreTestResult> {
  const { data } = await apiClient.post<BackupRestoreTestResult>(
    `/api/backup/backups/${encodeURIComponent(filename)}/test-restore`,
  )
  return data
}

export async function restoreBackup(
  filename: string,
  destinationId?: number,
): Promise<void> {
  const params = destinationId !== undefined ? { destination_id: destinationId } : {}
  await apiClient.post(`/api/backup/backups/${encodeURIComponent(filename)}/restore`, null, { params })
}

// ---------------------------------------------------------------------------
// API calls — OneDrive OAuth
// ---------------------------------------------------------------------------

export async function startOneDriveOAuth(): Promise<OneDriveOAuthStart> {
  const { data } = await apiClient.get<OneDriveOAuthStart>('/api/backup/oauth/onedrive/start')
  return data
}

export async function pollOneDriveOAuthStatus(): Promise<OneDriveOAuthStatus> {
  const { data } = await apiClient.get<OneDriveOAuthStatus>('/api/backup/oauth/onedrive/status')
  return data
}

File: backend/schemas/backup.py
"""Pydantic schemas for backup destinations and scheduling."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel

# ---------------------------------------------------------------------------
# Backup destination
# ---------------------------------------------------------------------------


class BackupDestinationRead(BaseModel):
    id: int
    name: str
    type: str
    enabled: bool
    rclone_remote_name: str
    rclone_config: str | None
    target_path: str
    created_at: datetime

    model_config = {"from_attributes": True}


class BackupDestinationCreate(BaseModel):
    name: str
    type: str  # local | smb | onedrive
    enabled: bool = True
    rclone_remote_name: str
    rclone_config: str | None = None  # JSON
    target_path: str = ""


class BackupDestinationUpdate(BaseModel):
    name: str | None = None
    type: str | None = None
    enabled: bool | None = None
    rclone_remote_name: str | None = None
    rclone_config: str | None = None
    target_path: str | None = None


# ---------------------------------------------------------------------------
# Backup schedule (stored in app_settings)
# ---------------------------------------------------------------------------


class BackupScheduleRead(BaseModel):
    enabled: bool
    schedule_type: str  # interval | cron
    interval_hours: int
    cron_expression: str | None
    include_uploads: bool
    notify_on_failure: bool

    model_config = {"from_attributes": True}

    @classmethod
    def from_settings(cls, s: Any) -> BackupScheduleRead:
        return cls(
            enabled=s.backup_enabled,
            schedule_type=s.backup_schedule_type,
            interval_hours=s.backup_interval_hours,
            cron_expression=s.backup_cron_expression,
            include_uploads=s.backup_include_uploads,
            notify_on_failure=s.backup_notify_on_failure,
        )


class BackupScheduleUpdate(BaseModel):
    enabled: bool | None = None
    schedule_type: str | None = None  # interval | cron
    interval_hours: int | None = None
    cron_expression: str | None = None
    include_uploads: bool | None = None
    notify_on_failure: bool | None = None


# ---------------------------------------------------------------------------
# Backup run status
# ---------------------------------------------------------------------------


class BackupDestinationResult(BaseModel):
    destination_id: int
    destination_name: str
    success: bool
    error: str | None = None


class BackupRunStatus(BaseModel):
    last_run_at: datetime | None
    last_run_status: str | None  # success | failure | None
    last_run_error: str | None = None
    destinations_results: list[BackupDestinationResult] = []


# ---------------------------------------------------------------------------
# Connection and restore test results
# ---------------------------------------------------------------------------


class BackupConnectionTestResult(BaseModel):
    success: bool
    message: str


class BackupRestoreTestResult(BaseModel):
    ok: bool
    integrity_check: str  # "ok" or SQLite integrity_check result
    tables_found: list[str] = []
    tables_missing: list[str] = []
    error: str | None = None


# ---------------------------------------------------------------------------
# OneDrive OAuth
# ---------------------------------------------------------------------------


class OneDriveOAuthStart(BaseModel):
    user_code: str        # e.g. "ABCD-1234" — user enters this at verification_uri
    verification_uri: str  # https://microsoft.com/devicelogin
    expires_in: int       # seconds until the code expires
    message: str          # human-readable instruction from Microsoft


class OneDriveOAuthStatus(BaseModel):
    done: bool
    token: str | None = None   # rclone_config JSON when done
    error: str | None = None
